Samples/2.3/FasterSample/src/Grains/FasterThreadPoolGrain.cs

```diff
@@ -0,0 +1,408 @@
+                session = lookup.StartSession();
+
+                var key = item.Key;
+                if (lookup.Upsert(ref key, ref item, Empty.Default, 0) == Status.ERROR)
+                {
+                    throw new ApplicationException();
+                }
+            }
+            finally
+            {
+                try
+                {
+                    if (session != Guid.Empty) lookup.StopSession();
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+        });
+
+        /// <summary>
+        /// Sets a range of item in the lookup.
+        /// This is a blind update.
+        /// </summary>
+        /// <param name="items">The items to set.</param>
+        /// <returns></returns>
+        public Task SetRangeAsync(ImmutableList<LookupItem> items) => Task.Run(async () =>
+        {
+            await semaphore.WaitAsync();
+
+            var session = Guid.Empty;
+            try
+            {
+                session = lookup.StartSession();
+                for (var i = 0; i < items.Count; ++i)
+                {
+                    var item = items[i];
+                    var key = item.Key;
+                    if (lookup.Upsert(ref key, ref item, Empty.Default, i) == Status.ERROR)
+                    {
+                        throw new ApplicationException();
+                    }
+                }
+            }
+            finally
+            {
+                try
+                {
+                    if (session != Guid.Empty) lookup.StopSession();
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+        });
+
+        public Task SnapshotAsync() => Task.Run(async () =>
+        {
+            await semaphore.WaitAsync();
+
+            var session = Guid.Empty;
+            try
+            {
+                session = lookup.StartSession();
+                if (!lookup.CompletePending(true)) throw new ApplicationException();
+                if (!lookup.TakeFullCheckpoint(out var token)) throw new ApplicationException();
+                if (!lookup.CompleteCheckpoint(true)) throw new ApplicationException();
+            }
+            finally
+            {
+                try
+                {
+                    if (session != Guid.Empty) lookup.StopSession();
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+        });
+
+        public Task<LookupItem> TryGetAsync(int key) => Task.Run(async () =>
+        {
+            await semaphore.WaitAsync();
+
+            var session = Guid.Empty;
+            try
+            {
+                session = lookup.StartSession();
+                LookupItem result = null;
+                if (lookup.Read(ref key, ref result, ref result, Empty.Default, 0) == Status.ERROR)
```

Comment:
A Read operation may return Status.PENDING, in which case we need to call CompletePending(true) in the same session/thread, either now or later on. If you call CompletePending(true) immediately after the Read, it will result in the operation forced to complete synchronously (the call will spin-wait for completion).

A different technique is to simulate the async/await model. To do this, you can associate the Read operation with a `TContext` of a `TaskCompletionSource<TOutput>` instead of `Empty`. Further, let the ReadCompletionCallback be responsible for setting this task as completed. You return this task to the user in the lookup path so they can await on it. We would still need to call CompletePending(false) on the session periodically to ensure that the I/O completions are processed and callbacks issued.

Reply:
### More on I/O and CompletePending

When a FASTER operation performs I/O, it returns immediately to the calling thread with a Status.PENDING. The asynchronous I/O completion callback (on the thread pool) eventually enqueues the operation result into a pending queue of completed operations, that is _local to the original session/thread_. The user is responsible for later calling `CompletePending` in order to drain this queue and issue the actual callbacks such as `ReadCompletionCallback`. This model allows threads to perform callback work locally on their own thread, and lets users control when the callback work occurs, e.g., after a batch of requests are issued.

Also, `CompletePending(false)` simply dequeues and processes callbacks for whatever operations have completed until that point - it does not wait. OTOH, `CompletePending(true)` spin-waits until all pending operations until that point are completed and dequeued by the thread.

Reply:
That makes sense, many thanks for the detailed explanation.